### PR TITLE
plugins/ccc: Add more settings

### PR DESCRIPTION
**Pull request by GaetanLepage**
Follow up of #1368 which I closed by accident.

### CODE REVIEW

File: plugins/utils/ccc.nix [hunk 1]
```diff
@@ -5,50 +5,218 @@
...
+        The default color used when a color cannot be picked. It must be HEX format.
       '';
 
+      inputs = mkOption {
+        type = listOfRawLua;
+        default = [
+          "ccc.input.rgb"
+          "ccc.input.hsl"
+          "ccc.input.cmyk"
+        ];
```
Comment (traxys) [inline plugins/utils/ccc.nix:33]: Why are we setting an explicit default? If this is important we should add a comment why

File: plugins/utils/ccc.nix [hunk 2]
```diff
@@ -5,50 +5,218 @@
...
+        default = [
+          "ccc.picker.hex"
+          "ccc.picker.css_rgb"
+          "ccc.picker.css_hsl"
+          "ccc.picker.css_hwb"
+          "ccc.picker.css_lab"
+          "ccc.picker.css_lch"
+          "ccc.picker.css_oklab"
+          "ccc.picker.css_oklch"
+        ];
```
Comment (traxys) [inline plugins/utils/ccc.nix:100]: Same

File: plugins/utils/ccc.nix [hunk 3]
```diff
@@ -5,50 +5,218 @@
...
+      };
+
+      outputs = mkOption {
+        type = listOfRawLua;
+        default = [
+          "ccc.output.hex"
+          "ccc.output.hex_short"
+          "ccc.output.css_rgb"
+          "ccc.output.css_hsl"
+        ];
```
Comment (traxys) [inline plugins/utils/ccc.nix:65]: Same

File: plugins/utils/ccc.nix [hunk 4]
```diff
@@ -5,50 +5,218 @@
...
+        '';
+      };
+
+      convert = mkOption {
+        type = listOfRawLua;
+        # TODO
+        default = [
+          "{ ccc.picker.hex, ccc.output.css_rgb }"
+          "{ ccc.picker.css_rgb, ccc.output.css_hsl }"
+          "{ ccc.picker.css_hsl, ccc.output.hex }"
```
Comment (traxys) [inline plugins/utils/ccc.nix:172]: I think this option should be a list of raw lua pairs, or at least a list of list of raw lua

Comment (traxys) [reply plugins/utils/ccc.nix:172]: It could also be an attr set, as this represents nicely pairings (where the keys & values would be raw lua)

**Review by traxys — CHANGES_REQUESTED** [review]

File: plugins/utils/ccc.nix (hunk 1 above)
Comment (MattSturgeon) [reply plugins/utils/ccc.nix:33]: [Looking upstream](https://github.com/uga-rosa/ccc.nvim/blob/f388f1981d222967c741fe9927edf9ba5fa3bcbe/doc/ccc.txt#L180-L203), that's the default used there.

My understanding is that we should be using a [default-null option](https://github.com/nix-community/nixvim/blob/c05fba47c733445098847399b61ddfd81ecbde55/lib/options.nix#L60-L72), which document upstream's default, as used when _our_ option is null. I.e. the `default` passed to `lib.mkOption` should be `null`.

Comment (traxys) [reply plugins/utils/ccc.nix:33]: Yes exactly

File: plugins/utils/ccc.nix (hunk 4 above)
Comment (GaetanLepage) [reply plugins/utils/ccc.nix]: The latter would not be RFC-42 style compliant, right ?
I went with the `listOf (listOf strLua)` approach.

File: plugins/utils/ccc.nix [hunk 5]
```diff
@@ -5,59 +5,252 @@
...
-        '';
-
-    lsp = helpers.defaultNullOpts.mkBool true ''
-      Whether to enable LSP support. The color information is updated in the
-      background and the result is used by `:CccPick` and highlighter.
-    '';
+      inputs = mkOption {
+        type = listOfRawLua;
+        apply = mapToRawLua;
+        default = [ ];
```
Comment (MattSturgeon) [inline plugins/utils/ccc.nix:29]: If this option is left as the default empty list, will an empty lua table be passed to `setup`? If so, will that override the plugins default `inputs`?

We would normally have the default be `null`, which (I assume) leads to the option being omitted from the lua table.

Also, if passing a value (that doesn't explicitly include the defaults) _unsets_ said defaults, we should probably make a note of that in the docs.

File: plugins/utils/ccc.nix [hunk 6]
```diff
@@ -5,59 +5,252 @@
...
+              "ccc.input.cmyk"
+            ]
+          ```
+        '';
+      };
+
+      outputs = mkOption {
+        type = listOfRawLua;
+        apply = mapToRawLua;
+        default = [ ];
```
Comment (MattSturgeon) [inline plugins/utils/ccc.nix:65]: Same question as `inputs`

File: plugins/utils/ccc.nix [hunk 7]
```diff
@@ -5,59 +5,252 @@
...
+              "ccc.output.css_hsl"
+            ]
+          ```
+        '';
+      };
+
+      pickers = mkOption {
+        type = listOfRawLua;
+        apply = mapToRawLua;
+        default = [ ];
```
Comment (MattSturgeon) [inline plugins/utils/ccc.nix:101]: Same question as `inputs`

**Review by MattSturgeon — COMMENTED** [review]
Looking good, just need to clarify whether some stuff should default  to `null` or if empty list is ok

File: plugins/utils/ccc.nix (hunk 5 above)
Comment (MattSturgeon) [reply plugins/utils/ccc.nix:29]: Is there a reason this can't use `helpers.defaultNullOpts.mkListOf`?

Comment (GaetanLepage) [reply plugins/utils/ccc.nix:29]: > If this option is left as the default empty list, will an empty lua table be passed to `setup`? If so, will that override the plugins default `inputs`?
> 
> We would normally have the default be `null`, which (I assume) leads to the option being omitted from the lua table.
> 
> Also, if passing a value (that doesn't explicitly include the defaults) _unsets_ said defaults, we should probably make a note of that in the docs.

Your concern is perfectly valid. But here, it is OK:
- Empty lists/attrs are indeed ignored by `toLuaObject`, so it has the same behavior as a `null` default.
- Now, why not `null` ? Well, just to simplify the `apply` function and not having to check if the value is `null` or a list.

Comment (GaetanLepage) [reply plugins/utils/ccc.nix:29]: With `mkListOf`, you cannot have an option where the user fills a list with lua strings that will get converted to raw lua automatically.

Comment (MattSturgeon) [reply plugins/utils/ccc.nix:29]: Thanks for explaining 👍

File: plugins/utils/ccc.nix [hunk 8]
```diff
@@ -5,59 +5,252 @@
...
+          List of color system to be activated.
+          `ccc-action-toggle_input_mode` toggles in this order.
+          The first one is the default used at the first startup.
+          Once activated, it will keep the previous input mode.
 
-    highlighter = {
-      auto_enable = helpers.defaultNullOpts.mkBool false ''
-        Whether to enable automatically on `BufEnter`.
-      '';
+          default:
```
Comment (MattSturgeon) [inline plugins/utils/ccc.nix:51]: This should updated to match #1530

```suggestion
          Plugin default:
```

File: plugins/utils/ccc.nix [hunk 9]
```diff
@@ -5,59 +5,252 @@
...
+          "ccc.output.hex.setup({ uppercase = true })"
+          "ccc.output.hex_short.setup({ uppercase = true })"
+        ];
+        description = ''
+          List of output format to be activated.
+          `ccc-action-toggle_ouotput_mode` toggles in this order.
+          The first one is the default used at the first startup.
+          Once activated, it will keep the previous output mode.
+
+          default:
```
Comment (MattSturgeon) [inline plugins/utils/ccc.nix:86]:
```suggestion
          Plugin default:
```

File: plugins/utils/ccc.nix [hunk 10]
```diff
@@ -5,59 +5,252 @@
...
+          "ccc.picker.css_lab"
+          "ccc.picker.css_lch"
+          "ccc.picker.css_oklab"
+          "ccc.picker.css_oklch"
+          "ccc.picker.css_name"
+        ];
+        description = ''
+          List of formats that can be detected by `:CccPick` to be activated.
+
+          default:
```
Comment (MattSturgeon) [inline plugins/utils/ccc.nix:116]:
```suggestion
          Plugin default:
```

File: plugins/utils/ccc.nix [hunk 11]
```diff
@@ -5,59 +5,252 @@
...
+            "ccc.picker.css_oklch"
+            "ccc.output.hex"
+          ]
+        ];
+        description = ''
+          Specify the correspondence between picker and output.
+          The default setting converts the color to `css_rgb` if it is in hex format, to `css_hsl`
+          if it is in `css_rgb` format, and to hex if it is in `css_hsl` format.
+
+          default:
```
Comment (MattSturgeon) [inline plugins/utils/ccc.nix:209]:
```suggestion
          Plugin default:
```

**Review by MattSturgeon — CHANGES_REQUESTED** [review]
After a quick scan, everything seems resolved. Just need to update some manually crafted descriptions to match #1530.

**Review by MattSturgeon — APPROVED** [review]

**Comment by GaetanLepage** [comment]
@traxys we need your approval to merge ;)

**Review by traxys — APPROVED** [review]

**Pull request merged**